### ISSUE TITLE
Float.round is inconsistent

### ISSUE THREAD

**Issue by ewildgoose** [elixir-lang/elixir#3400]
The rounding from Float.round is inconsistent, eg

iex(121)> Kernel.round(1.255 * 100) / 100
1.25

iex(123)> Kernel.round(125.5) / 100
1.26

iex> Kernel.round(Kernel.round(1.255 * 1000) / 10) / 100
1.26

The core issue is numerical instability due to the multiplication by a power of 10 (which itself has to be encoded as an approximation in floating point format)

The following *appears* to correctly implement the same rounding function as Kernel.round (ie round away from zero) without instability.  However, it would be sensible to include a test harness for very large/small numbers and also floats in "denomalised" form.

```
  @spec round(float, 0..15) :: float
  def round(number, precision \\ 0) when is_float(number) and precision in 0..15 do
    power = power_of_10(precision)
    power2 = power_of_10(precision+1)
    # Avoid numerical instability by pre-rounding to one extra digit of precision
    Kernel.round(Kernel.round(number * power2) / 10) / power
  end

  Enum.reduce 0..21, 1, fn x, acc ->
    defp power_of_10(unquote(x)), do: unquote(acc)
    acc * 10
  end
```

**Comment by Kabie** [comment]
This behavior exists in many languages. I think people should use [decimal](https://github.com/ericmj/decimal) when they do mind.

**Comment by bitwalker** [comment]
@Kabie that's not a good reason to leave it unaddressed if there is a better solution available in my opinion.

**Comment by edescourtis** [comment]
You could consider using a  rational numbers library like this one https://github.com/fraczak/rat .

Not sure what is available for Elixir directly but this should work.

**Comment by ewildgoose** [comment]
Hi all. There is some discussion on this on the elixir-talk mailing list.  I have struggled to show the core problem in a succinct way, so can I refer you to the discussion there?  I agree entirely about using a dedicated datatype to work in a specific number base, but *this* bug report is about an inconsistency between two core-lib functions which implement a certain numerical function (round).

**Comment by josevalim** [comment]
Folks, regardless if some problem can be solved in some other way, all
float operations should be as correct and precise as possible. This is not
the place to discuss using a rational or decimal library because they won't
change the fact round could be better implemented.

-- 


*José Valim*
www.plataformatec.com.br
Skype: jv.ptec
Founder and Director of R&D

**Comment by wishdev** [comment]
round is 100% correct - the issue is simple - 1.255*1000/10 != 1.255*100 - that's the way floating point numbers work sometimes. Round can't fix issues it has no control over.

**Comment by ericmj** [comment]
@wishdev Please see the discussion on the mailing list and try the proposed algorithm. We can demonstrate that the new algorithm improves the examples @ewildgoose provides. If you can find examples where the new algorithm performs worse than the current then please show counter examples.

**Comment by wishdev** [comment]
Ok

Float.round(1.249, 1) = 1.2

If I read your proposal correctly you offer the following

Kernel.round(Kernel.round(1.249 * 100) /10)/10 = 1.3

The error exists on anything from 1.245 < x < 1.25 or most variant thereof. Obviously it's not that simple because rounding "issues" exist all over the place with floating numbers.

**Comment by ewildgoose** [comment]
Oops, yeah, busted.   I also think that Float.ceil / Float.floor are affected since they use the same underlying algorithm?

I have done a google search for some ideas, but I didn't turn up anything which correctly addresses this within an IEEE double.  

I think for a perfectly correct answer we can convert to a string of digits, perform the rounding and convert back again (effectively work in the decimal domain).  At present this can't be done using Float.to_string(1.249, [decimals: 1, compact: true]) because the rounding isn't quite as we desire.  However, I have a separate issue open for a different algorithm for the to_string implementation. Coupling these might give a solution?

Any other ideas?

**Comment by ewildgoose** [comment]
I have implemented a possible new Float.to_string algorithm. This allows optimal output with flexible rounding to arbitrary precision.  Its feasible to use this to implement round/floor/ceil for decimal values to arbitrary precision.

I haven't found any papers on doing perfect decimal operations whilst staying in IEEE double space. I think to properly implement arbitrary decimal rounding will require working in decimal.  The code here achieves that goal
* https://github.com/ewildgoose/elixir-float_pp/blob/master/test/rounding_test.exs

I suggest moving discussion to issue #3413

**Comment by josevalim** [comment]
Let's keep the discussion here because we should not mix the rounding discussion with the to_string one. Maybe the solution will be the same but they are definitely separate issues.

**Comment by josevalim** [comment]
> I haven't found any papers on doing perfect decimal operations whilst staying in IEEE double space

Yes, because we cannot expect perfect decimal operations on IEEE double space. If the solution is to convert to float -> decimal -> perform operations -> back to float then it is not an appropriate solution. You are going to have rounding errors with floats, making it work for 1.2 is not adding much given that it still won't work for 1.1 + 0.1. We should be as precise as possible without resorting to decimal conversion.

**Comment by josevalim** [comment]
In other words, round is an operation on a float, it is not an operation on a decimal. As adding a number to a float can introduce imprecision, rounding a float may introduce imprecision too. We can reduce the amount of imprecision, we won't, however, guarantee there is no imprecision.

**Comment by ewildgoose** [comment]
* Currently the rules on how Float.round should operate aren't specified.  A lot of misunderstanding can be cleared away if we *specify* what this function should output for a given input...
* I propose that we specify Float.round to operate as "Directed rounding to nearest, ties round away from zero", which is most similar to how Kernel.round work and how it nearly works right now
** https://en.wikipedia.org/wiki/IEEE_floating_point#Rounding_rules
* Once we agree on a specification we can then test our implementation for correctness. 
* I assert that the current implementation doesn't meet this specification (in fact it doesn't meet any common rounding specification). 
* I think we all agree on the underlying reason we don't meet the spec, and that is of course because the current implementation relies on doing a number of floating point operations *first* before applying the rounding. Those operations introduce an error, which affects the accuracy of the final output.  We either need to find a way to eliminate the float operations or analyse the operations and correct for the introduced error.


P.S. I would personally prefer rounding to be specified as "Directed rounding to nearest, ties to even", but that seems a long way from where we are now.  However, its worth considering that for many requirements, this function (if specified as above) won't be useful and might be giving unexpect

**Comment by romul** [comment]
I agree with @wishdev. You couldn't detect what is passed to round 
`1.255 * 100` or `125.49999999999999` or `125.5 - 1.0e-14`
So, trying to fix float numbers you would introduce the issue in the cases with correct values.

**Comment by ewildgoose** [comment]
Hi, actually, could you please look VERY closely at the original report before responding further? (I mean this respectfully, quite possibly you already have)  This thread and the related one have produced lots and lots of responses shooting from the hip and the problem is a subtle one

I think perhaps my original report is too subtle? It looks as though the arithmetic (*100, etc) is important, but in fact its the opposite, the report shows HOW elixir is internally computing the rounding. So that you get several different answers highlights how the internal algorithm in elixir *relies* on IEEE floating point to do unexpected things. 

So yes, you cannot detect whether you get passed x*100 or whatever, that's not the point. The point is that the algorithm is implemented internally by doing this *100 stuff internally, hence tickling the rounding issues that we all are aware of.  I'm trying to show how the internals cause a problem, not how the input matters.

I dispute that randomly rounding the way a human would round (and sometimes not) is "principle of least astonishment"?  

Note, this is the point, the rounding is random from the point of view of base10.  I accept this might not be clear in the original report, but my claim this is a "bug" is because the rounding is inconsistent, not that IEEE can completely represent all floating point numbers in base 10. ie I claim there must exist a different algorithm which WOULD produce "least astonishment", eg compare the results with the C library to see the difference...

Also note there is more than one definition of "rounding".... So if you see my previous comment, FIRST we need to specify what the rounding rules SHOULD be, then we can see if the implementation meets that spec...  For now be careful using elixir floats for base10 type operations, as there are some significant differences to how other libraries will operate.

**Comment by squallified** [comment]
>Note, this is the point, the rounding is random from the point of view of base10. I accept this might not be clear in the original report, but my claim this is a "bug" is because the rounding is inconsistent, not that IEEE can completely represent all floating point numbers in base 10. ie I claim there must exist a different algorithm which WOULD produce "least astonishment", eg compare the results with the C library to see the difference...

I don't think there is another way to round IEEE 754 floats (except converting them to something else). Even languages that provide the option to supply the number of decimal places use the same method: [.NET round implementation](http://referencesource.microsoft.com/#mscorlib/system/math.cs,88)

**Comment by josevalim** [comment]
Thanks everyone.

After reviewing the whole discussion a couple times and collecting feedback from other languages, I don't believe this is an issue. @wishdev puts it really well, we will have precision issues always in some interval, regardless of how we implement it. I also don't agree rounding behaviours are inconsistent: they are rounding at different precisions, which means precision differences will show up.

Finally, I don't think the solution should be to convert to a string and work on that. The whole point of using float numbers is so operations are fast (albeit imprecise). If we implementing Float.round differently, it will still be imprecise (because inputs and outputs are still floats, which are imprecise) but slow. If you need precision, then absolutely do not use float.

**Comment by awetzel** [comment]
sorry to ping this issue again, but I have just had a big bug very difficult to debug because of this function, and I think that your first impression @josevalim is the one of any dev using the function, and so maybe afterward thinking is too much thinking :) . 

> they won't change the fact round could be better implemented

My point is pretty close to the one of @ewildgoose. ":erlang.round" function is determinist, since the result is an integer, so useful. But the `Float.round` function is not deterministic  because it uses a float computation to get the result. But if your want to use a "non-deterministic" float, then in which use case is it necessary to "round it" ? just keep your float as is. I think this is the reason why there is no such function in OTP, and it makes sense. If you round a number, you want precisely to transform a fuzzy number in a meaningful number with a given accuracy, at any cost. 

When you do computation then the concept of *imprecision* is meaningful, when you are converting/transforming a value (round), then the concept itself of *precision* has no meaning, : but the transformation should be deterministic. 
Here, applying an "imprecise" computation (which could be useful) to obtain a functional transformation (rounding), elixir actually creates a non deterministic transformation (which is useless).
I think it introduces a useless (if you understand what it does) or dangerous (if not) computation, right into the std lib. If the user / dev understands the lib, I do not think about any use case where this function could be used.

Still I think usage of `%Decimal{}` is not an answer to this particular issue : 
- float calculation is often use, because it is simpler, more efficient for computation, and most of the time sufficient to get a job done
- float are everywhere in many libraries, for instance coding/encoding libs, and legitimately, because the overhead of a "decimal" structure creates often an unnecessary complexity to handle the values directly with bifs
-  BUT float calculation are rarely useful if you cannot use in the end the result in a problem which needs "determinism". For that,  erlang implement the right algorithm in C in the NIF (`float_to_binary` and in erlang in io_lib format `io_lib_format.fwrite_g`).

Since `:erlang.float_to_binary` already implements the right rounding, it does the job to make a float deterministic, so IMHO I would have prefer that : 
- ever the entire function should be deprecated then removed with a warning
- else, use the determinist and not so inefficient `float |> Float.to_string(decimals: 2) |> String.to_float` since it uses 2 NIF bifs, with an explicit doc explaning it is not a float computation result, so a lot less efficient.
- if you do not agree with above points : at least put a warning in the doc, because currently, the doc is very misleading : even the examples seems to tell you that the function does a "half up rounding" (`Float.round(5.5674, 3)=> 5.567`, `Float.round(5.5675, 3)=>5.568`) which is wrong. Even worst ! it is true most of the time.

Have a great day :)

**Comment by awetzel** [comment]
and @josevalim , I do not agree with 

> If we implementing Float.round differently, it will still be imprecise (because inputs and outputs are still floats, which are imprecise) 

compute the mantissa exponent from the binary float representation of the system, an make rounding accordingly (see `io_lib_format.fwrite_g` for erlang impl in `erts/emulator/sys/common/erl_sys_common_misc.c` `sys_double_to_chars_fast` which are the impls used by `:io.format('~0.2f')` and `:erlang.float_to_binary`, to get digits are not an imprecise computation as the current one at all.

**Comment by awetzel** [comment]
Sorry to "ping" this issue again, but I quite agree with @ewildgoose , Float.round is missleading : 
- because "round" has a meaning for user/dev/math, and the meaning is a functional, determinist transformation from a "real" and non-deterministic result. If you do not want determinism, then you do not round, so the current non-determinist function has no use case (except if you think the function does something it does not).
- because de doc is missleading the examples : 

```
iex> Float.round(5.5674, 3)
5.567
iex> Float.round(5.5675, 3)
5.568
```

tells the user that the function is doing a "half_up" rounding, which is actually not the case. Worst : it is the case most of the time ! 

So I advocate either the removal/ deprecation of the function, or at least a big warning in the doc.

Finally I do not agree @josevalim when you say that  

> If we implementing Float.round differently, it will still be imprecise

error rounding can be computed with the mantissa exponent, which is done when you use `:erlang.float_to_binary` `erts/emulator/sys/common/erl_sys_common_misc.c` and iolib `:io_lib_format.fwrite_g`. So it is possible to get a deterministic result both way, but still it is not in both cases a "good" result because : 
- float_to_binary mitigate the "floating error" but doing that, it does a non-mathematical rounding : `Float.to_string(3.254555599,decimals: 2)  === 3.26`
- `fwrite_g` does the mathematical rounding, but based on all the digits deduced from the "float", so it does not handle the "floating error" if the float is the result of a floating point computation.

But both are determinist so are useful ! `Float.round` is not. Using both functions together, you can create a mathematically correct float rounding.  

A POC there : https://gist.github.com/awetzel/dc0fa03f7b280642f3bbf7113be3ee73

Also, currently `%Decimal{}` library is not mature enough, not able yet today to get correct rounding, so it is a missing important feature un Elixir. And it has it place into stdlib.

Have a good day :)

**Comment by josevalim** [comment]
Thank you @awetzel for the detailed reply! I have performance concerns about using an implementation that relies on the conversion from and to binary. Should we instead calculate the mantissa ourselves and do the rounding from there?

**Comment by wishdev** [comment]
What about 

DecimalFloat.round(0.100000000005 + 0.2, 14) - 0.3 from the POC
vs
Float.round(0.100000000005 + 0.2, 14) - 0.300000000005 from the system

Could we please see at least a single example where this works better without adding arbitrary limitations to Float.round - i.e apparently only allowing 7 digits of precision.

**Comment by awetzel** [comment]
first : the POC works only with precision < 7 because I didn't need more when I did it :) and is not arbitrary, it is because the floating point error IS `< 0.0000001` and rounding is very rarely needed above "10^-7", and it is a POC, it should be possible to compute the max diff with the mantissa exponent, but it is not a simple task :) .

I think your question is not a use case I can identify. When you round : you want to get a deterministic result from a result. so here, adding float-point calculation to the result, by definition, re-includes non-determinism (the float binary structure is deterministic, decimal computation on it is not). It makes sense to apply a round to a floating point computation result, not the other way.
 
Also : the POC function had a limitation of 7 decimal precision, but with this limitation the function could be useful for users who want to round with less than 7 decimals, . 
The current `Float.round` function is not usable at all, because when you "round", you are not doing floating point computation, you are trying to do a deterministic transformation. Could you give me an example of use case where a random rounding is acceptable ? 

I proposed a solution to make the point that something is possible. It is not a complete solution. 

My point is also that if no solution is implemented, then I advocate a function deprecation. You can say that all users doing floating point computation and still want somme determinism are not "right", but this actually make sense in many case. And the random characteristic of `Float.round` is not explicit at all.  And if the function was removed, then these users/devs could use deterministic `Kernel.round` to make a "floating - point computation  to deterministic result", and they will do it. If you keep the function as is, then you are missleading them.

**Comment by awetzel** [comment]
also I have a file with 1Go file of floating points computations with a "2" decimal point rounding, which is my personal use case : and had : 
- 100% of mathematically correct result with the POC function
- ~10% of errors with `Float.round`
- ~30% of errors with `String.to_float`
I think that makes the point that something is doable.

**Comment by wishdev** [comment]
Again not a single actual example provided - you dismiss my actual example and again show not a single example that you fix. 

Also since when does the entire rounding concept end at 2 or 7 decimal points? 

You use large words without understanding what they actually mean - because you have never and can never show a single example where Float.round(x, y) returns a different result for the same x and y - which is what non-deterministic means. It does not mean that because floating point computations are inaccurate (i.e 0.1 + 0.2 != 0.3) that a rounding function becomes "non-deterministic" In other words Float.round(0.3, y) may not equal Float.round(0.1 + 0.2, y) because the x in both calls are not the same.

It's nice to color over something and make it look pretty (i.e. locking down to 7 decimal points)  - but before you attack something you should at least provide some example of what you actually want to fix.

**Comment by wishdev** [comment]
And my mistake from the example above

DecimalFloat.round(0.100000000005 + 0.2, 14) == 0.3 from the POC
vs
Float.round(0.100000000005 + 0.2, 14) == 0.300000000005 from the system

**Comment by awetzel** [comment]
Ok @wishdev , I actually don't care if you use this tone :), I am a bit overwhelmed by my work currently, so I do not want anyone to use this solution. This function broke a service with million of user, so I thought that it was useful to share. 
Have a good day.
I do not attack anything really :)

**Comment by wishdev** [comment]
I would love to see something that worked better - but 4 responses in and you have provided still no actual example. Not a single line of code that Float.round did incorrectly - not a single example at all. Why shouldn't one be rather skeptical when the simplest of things seems not to be forthcoming.

**Comment by josevalim** [comment]
To avoid the discussion from going off-topic, I will close the discussion and open up a new issue later on.

**Comment by josevalim** [comment]
Hello everyone,

I have done some research on this topic and I will present my conclusions.

IEEE 754 is a standard for floating pointing arithmetic. It defines operations in two radices, 2 and 10. The standard defines what is correct rounding:

> correct rounding: This standard’s method of converting an infinitely precise result to a floating point number, as determined by the applicable rounding direction. A floating-point number so obtained is said to be correctly rounded

To avoid confusion, I won't use the term "rounding" on its own. I will use the term  "rounding precision" to refer to the rounding precision and "rounding direction" to refer to the direction (nearest even, ceil or floor), as per the standard. Elixir today supports nearest even (via Floor.round), ceil (via Floor.ceil) and floor (via Floor.floor). 

The standard also specifies that, when reducing the precision of a float, rounding must be applied:

> If the conversion is to a format in a different radix or to a narrower precision in the same radix, the result shall be rounded as specified in Clause 4. Conversion to a format with the same radix but
wider precision and range is always exact.

The current implementation in Elixir has no proof. So we will assume it is incorrect. It also seems to violate the last sentence above: it is not guaranteed our conversion is exact when increasing the precision (i.e. the precision argument given to `Float.round` is more than the precision we have on the float).

While the standard provides a definition of rounding, it does not say how to perform rounding algorithmically. Nor it says how to convert between radices besides requiring the conversion should be correct. However, the standard is clear that conversion between radices inside operations should be discouraged or disallowed:

> Language standards should disallow, or provide warnings for, mixed-format operations that would cause implicit conversion that might change operand values.

This means that the suggestions proposed above to "fix" Float.round are invalid according to the specification because they first convert a float from radix 2 to radix 10!

Luckily there are papers that define how to correctly round binary and decimal floats, back by proofs:

  * [Correctly Rounded Binary-Decimal and Decimal-Binary Conversions](http://ampl.com/REFS/rounding.pdf)

  * [Printing Floating-Point Numbers Quickly and Accurately with
Integers](https://333341ac-a-698ebdfa-s-sites.googlegroups.com/a/loitsch.com/florian/publications/dtoa-pldi2010.pdf?attachauth=ANoY7crKuCjCMB1AbdWipAWks-oEEAFNolAiqowu2WXD1hCjylmz_DeaNExvof4mVyn6hUiNssr928yGQZ_AxJWcXb4yurXAW2qVBYKMk4qHU8qCSQjc-TTQpzsFRos-fSLBab5FDM6U3YwmsoUE7xIK2qhUUl8SxjyXGKRPwCcIT077o9uezXDep5AOnxtVoiZm57-q-0R3JJFZwYDgyCbRSqXw9s_27eK6L4K9N3VUtHMopkOygAw%3D&attredirects=0)

Both papers also provide reference implementations so we can actually evaluate our implementation against. When trying those implementations, rounding to the nearest even, still has "surprising" results such as:

    round(12.3415, 3)
    12.341

However, those results are not "surprising". The number 12.3415 is effectively represented as 12.34149999 and therefore **correctly** rounds to 12.341 when using binary floating point operations. **This result is according to both the standard and literature**. If your implementation returns 12.342, it is not a rounding operation on a binary floating point.

Proposals above, such as DecimalFloat, are effectively converting the float to decimal base and then back to a binary float, which is discouraged by the standard. If you want to perform a decimal rounding operation on a float, you should explicitly convert it to a decimal. Converting the binary float to a decimal behind the scenes to perform rounding would be akin to converting floats to decimals before performing an addition.

To move forward, we will add a correct implementation of binary floating rounding in Elixir. We will not support decimal floating rounding, if you desire that, you must explicitly convert it to a decimal according to the standard.

I will also reopen the discussion. Because the discussion so far was based on "impressions", if you disagree with the proposals or the results above, please include:

  * A quote or a section in the IEEE 754 standard that backs up your claims
  * A paper or reference implementation, similar to the ones linked above, that backs up your results

The discussion will be locked again if it moves into arguments that have not been formalized or cannot be backed up by the standard.

Thanks everyone for your patience.

**Comment by ewildgoose** [comment]
Jose, thanks for phrasing the issues succinctly. 

There is clearly no perfect answer, here, it's about definitions and intended purposes. However, for many purposes I would find it more useful to get the following result, disregarding that IEEE 754 would indicate otherwise

```
round(12.3415, 3)
12.342
```

As "justification by example", I examined LibreOffice, Excel and Java and their preference is the result above (screenshots can be provided). Note: For the benefit of others reading this, there are plenty of counter examples in other products.

**Comment by josevalim** [comment]
Since Erlang implements IEEE754, with all operations such as float additions, subtraction and so on based on IEEE754, I believe it is important to follow IEEE754 recommendations.

The example you posted above is inconclusive as proof that those languages are operating on radix 10. In fact what you showed is that those languages perform rounding half up instead of half even. This seems to be the case with Java:

```java
java> String.format("%.5g%n", 12.3425)
java.lang.String res12 = "12.343"
java> String.format("%.5g%n", 12.3415)
java.lang.String res13 = "12.342"
java> String.format("%.5g%n", 12.3405)
java.lang.String res14 = "12.341"
```

It also seems to be the case for Excel. So based on those examples alone, it seems they also follow IEEE754 and are operating on radix 2 albeit they use another rounding technique.

It seems `Kernel.round/1` in Elixir also rounds half up so Float.round/2 should likely do the same.

**Comment by josevalim** [comment]
Please disregard my previous answer. Even with a half up direction, those numbers above should have been rounded down since 12.3415 is effectively 12.3414999999999999147. I will have to investigate how Java and Excel are performing rounding to give an appropriate answer.

**Comment by ewildgoose** [comment]
Are these results as you would expect?

```
iex> FloatNew.round(12.341444444444442, 15)
12.34144444444444
iex> FloatNew.round(12.341444444444443, 15)
12.341444444444445
iex> FloatNew.round(12.341444444444445, 15)
12.341444444444445
iex> FloatNew.round(12.341444444444446, 15)
12.341444444444448
iex> FloatNew.round(12.341444444444447, 15)
12.341444444444448
iex> FloatNew.round(12.341444444444449, 15)
12.341444444444448
```

**Comment by awetzel** [comment]
First thank you so much for your quick and hard work Jose : Float.ratio, new function comments, and determinist fixed-point rounding/dec conversion close the main issue.

But I still think a combination of fixed-point rounding->decimal conversion then decimal "digit rounding" would be a nice to have, but actually it makes more sense in a "Float to String" or "Float to `%Decimal{}` conversion than in a `Float.round` conversion.

**Comment by josevalim** [comment]
@ewildgoose thank you, there was loss of precision in the division operator we were using at the end of rounding, I am now encoding the integer to float back without loss.

**Comment by ewildgoose** [comment]
It would seem useful to support rounding to take a negative precision (ie round to hundreds, thousands). 

Also it would seem desirable to support precision the full size of float, ie much more than 0..15. Consider that you might want to round 1.2345e-50 to some number of DP. I would have some cause to ponder whether there is a loss of precision on the final divide doing this, but I wonder if your recent change would manage this?

**Comment by josevalim** [comment]
@ewildgoose those are feature requests. The goal of this issue is to fix the precision bug in the previous implementation.

If you need more features in Float.round, please start a new discussion in elixir-lang-core. But I would rather focus on improving the performance of this function using more recent algorithms before adding more features into it. So it is likely new features are releases until a faster algorithm is in place.

With the recent changes we should have no loss in precision.

**Comment by ewildgoose** [comment]
OK, so fastest performance is to use a NIF.  I observe for example that io_lib.format implements a function which passes your current test cases:

```
def jose_round(number, precision), do: :io_lib.format("~.#{:precision}f", [number]) |> hd |> :erlang.list_to_float
```

I concede I didn't dig through the source to find why it's not faster, but it seems like it should be fixable? (currently it's slower than your proposed replacement function)


```
## Bench
benchmark name        iterations   average time
using OldFloat.round      100000000   0.07 µs/op
using FloatNew.round      500000   3.55 µs/op
jose_round                500000   3.87 µs/op
```

I also present the following function for analysis. 

``` 
def ed_round(number, precision), do: :erlang.float_to_binary(number, [{:decimals, precision}, :compact]) |> :erlang.binary_to_float
```

I claim there is no one true way to implement a round() function, it's simply about defining how you *want* it to work. Therefore referring to my specification points I gave you previously, this function differs in that point 3) it's intended for use by humans. It (empirically) implements a different rounding philosophy, so doesn't meet your specification, but:

- It's algorithm is preferable to *me* (in many cases)
- It's unexpectedly faster...

```
## Bench
benchmark name        iterations   average time
using OldFloat.round      100000000   0.07 µs/op
ed_round                10000000   0.19 µs/op
using FloatNew.round      500000   3.55 µs/op
jose_round                500000   3.87 µs/op
```

Perhaps there is some mileage here to make use of the io_lib method at full speed AND/OR offer both rounding methods as options. Note, I concede these are early results, I didn't (yet) build a full test case to explore potential edge conditions.

**Comment by josevalim** [comment]
Thank you @ewildgoose. Improving the performance of Float.round is always welcome.

However, as previously asked, let's please deter from moving the discussion to solutions that are not backed up by reference implementations or academically:

> I claim there is no one true way to implement a round() function, it's simply about defining how you want it to work. 

As pointed earlier, we have papers and algorithms, as well as the IEEE754 standard, that we have based our current results on. Also, as previously stated, your implementation (or Java's) are also correct. However they perform the conversion on decimal value (after a free-point conversion) while we have chosen to use a fixed-point conversion in Elixir rounding. We choose this behaviour because we believe this to be inline with IEEE754 recommendations.

I am open to other solutions if they can be backed by either the standard, academy or industry. If the discussion moves to personal claims, I will lock the conversation again.

**Comment by ewildgoose** [comment]
I answer with care, because you sometimes assume I'm being bloody minded when I'm just trying to be explicit...

Because I don't fully understand your reply here, therefore others might not, let me try and answer in several ways, hopefully clarifying several questions others might have. (please forgive if I mis-answer though - I am *trying* to answer what I think is your point)

Firstly, I offered the implementation above (jose_round) BECAUSE as near as I can tell it's LITERALLY the same implementation that you now use?  Your current implementation in pseudo code (appears to me) to be equivalent to:
- Generate long sequence of digits equivalent to the exact decimal representation of the float number
- These digits are generated as an integer numerator and decimal divisor (ie 10^n)
- Truncate the numerator digits, leaving one digit extra for final rounding
- Apply rounding to the final (extra) digit, leaving the requested number of digits
- Do a divide by 10^n to convert the int back to a float number

Hence why I looked for normal string printing, since your algorithm should give the same results to simply printing the precise string out, then truncating and rounding (which is why I don't understand why you claim to be doing this in base2 or whatever you mean by "use a fixed point conversion"? Your algorithm LITERALLY converts to decimal and rounds the resulting decimal number using normal decimal rounding rules?)

eg to illustrate the equivalence, using as input, the decimal value 5.5675, this will be converted in the (new) round/2 function to the following list of digits (stored as an arbitrary length integer), this happens just after the line with the comment "# Convert the numerand to decimal base"

`556749999999999989341858963598497211933135986328125`

...which we can compare to
```
iex> :erlang.float_to_binary(5.5675, [{:decimals, 50}, :compact])
"5.56749999999999989341858963598497211933135986328125"
```

So, to avoid misunderstanding. My point is not that one implementation is better, simply that I claim your algorithm is completely equivalent to doing a "naive" decimal print output on the given floating point number, then truncating and rounding those decimal digits to the required precision.  ie there is no need to reference any papers, it's just a straightforward conversion and truncation+rounding of the literal decimal string. Nothing wrong with this at all!

The only reason I offered an implementation converting to string output and back to float is that it may be possible to use built in NIF functions, hence doing it much faster, despite theoretically being a slower method (if we were implementing say in C, your current method should be faster, in Erlang, perhaps not...)


Now, if I can rule off the above. Just to agree with you. As near as I understand, the IEEE754 conversion rules state something to the effect of: "assume the input is computed to infinite precision, now do your conversion on that infinite precision number and voila".  Hence why inputs like 1.50, are represented as an interval in IEEE754, the centre of that interval being 1.499999999...(etc), which under very literal interpretation of the IEEE recommendations is closer to 1.0 than 2.0.

My comment to you about "defining how it works" seems to be getting misunderstood because you are mapping the word "round" to a definition in the IEEE document... How about I call my function something different?  Lets call it "make_it_so/2".  OK, now we are free to implement this function however we wish?  Hopefully now you don't feel I'm disagreeing with you when I state a separate need for such a function?

So for sure, call it a feature request, I will keep this brief for that reason, but my point boils down to: sometimes it's useful to have a function where 
`  floor(1.5, 1) == 1.4`
 and sometimes it's useful to have a function where 
`  floor(1.5,1) == 1.5`
Neither are "wrong", eg IEEE754 would prefer the first one. However, my government requires the latter (when I do my tax returns)...

Anyway, since you want that to be a feature request, fine. However, please consider the previous comment since I think there is likely a higher speed implementation available through already existing Erlang NIF functions?

**Comment by josevalim** [comment]
@ewildgoose I just want to say I agree with everything you said.

> How about I call my function something different? Lets call it "make_it_so/2".

Yes, this will help move the discussion forward. For reasons you summed up really well, I would like the function named `round` to adhere to IEEE754. However, if you want to introduce a function named `make_it_so/2` or `decimal_round/2`, we can start such discussion. I would suggest for us to move it to a separate issue to keep things organized.

Any ideas to improve our current implementation is also welcome and we can continue to do those in this thread.

> (which is why I don't understand why you claim to be doing this in base2 or whatever you mean by "use a fixed point conversion"? Your algorithm LITERALLY converts to decimal and rounds the resulting decimal number using normal decimal rounding rules?)

The papers linked above (and others) describe two ways of performing conversion from binary to decimals: free-form and fixed point.

On a free-form conversion, you convert the whole binary to a decimal with an optimal (minimum) amount of digits to guarantee a conversion back to the same float. When you round on top of a free-form calculation, you are rounding on top of this optimal decimal approximation.

The fixed-point one converts the float to a decimal with a fixed precision. Therefore, when the rounding happens, it is happening on top of actual, non-approximated float value, as it is converted to decimal. While the current Float.round implementation does convert the binary to a decimal base, it is an exact representation of the float in decimal base and not the approximated decimal one.

In other words, `ed_round` is rounding on `5.5675`, which is the optimal decimal representation of `5.56749999999999989342`. `jose_round` is rounding on the float number itself. The fact both are in decimal base does not matter, what matters is if an approximation was performed as part of the process to convert to decimal base or not.

**Comment by Mysql-x-devapi-hind** [comment]
> Let's keep the discussion here because we should not mix the rounding discussion with the to_string one. Maybe the solution will be the same but they are definitely separate issues.





https://stackdev.space/questions/how-accurate-is-the-comparable-really-compared-to-its-initial-appearance
